impl(bigquery): v2 job initial files

jsrinnn opened the pull request:
Initial mostly empty files for the Bigquery V2 job API.

For details please see design doc: go/bq-odbc-client-design

I have run checkers-pr, clang-tidy-pr for bigquery. Running it from root was taking a very long time even the second time around.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10733)
<!-- Reviewable:end -->

coryan commented:
/gcbrun

devbww commented on google/cloud/bigquery/v2/minimal/internal/job_idempotency_policy.cc:
```diff
@@ -0,0 +1,28 @@
+
+namespace google {
+namespace cloud {
+namespace bigquery_v2_minimal {
+namespace internal {
```
We prefer `bigquery_v2_minimal_internal` these days.

devbww commented on google/cloud/bigquery/v2/minimal/internal/job_idempotency_policy.cc:
```diff
@@ -0,0 +1,28 @@
+}  // namespace internal
+}  // namespace bigquery_v2_minimal
+}  // namespace cloud
+}  // namespace google
```
Add newlines at the end of all the files.

coryan commented on google/cloud/bigquery/v2/minimal/internal/job_idempotency_policy.cc:
```diff
@@ -0,0 +1,28 @@
+// Copyright 2021 Google LLC
```
nit: 2023, here and anywhere else you add a copyright notice.

coryan commented on google/cloud/bigquery/v2/minimal/internal/job_idempotency_policy.h:
```diff
@@ -0,0 +1,31 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Internal interface for Bigquery V2 Job resource.
```
I am not sure we want to reference internal source files here.  We can, but seems a bit odd.  And generally we document the C++ symbols more than the files.

jsrinnn commented:
I have addressed all feedback in the latest commit. PTAL.

devbww commented on google/cloud/bigquery/v2/minimal/internal/job_idempotency_policy.cc:
```diff
@@ -13,16 +13,13 @@
-    
-}  // namespace internal
-}  // namespace bigquery_v2_minimal
+namespace bigquery_v2_minimal_internal {
```
PS: We should probably be adding `GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN` here.

jsrinnn replied:
Okay. I'll add it similar to the following:

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/internal/bigquery_read_connection_impl.h#L38

and

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/internal/bigquery_read_connection_impl.h#L104

jsrinnn replied:
Followup question: I also notice the following in other header files towards the end of the namespace

namespace gcpcxxV1 = GOOGLE_CLOUD_CPP_NS;  // NOLINT(misc-unused-alias-decls)

Reference: https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/internal/bigquery_read_connection_impl.h#L105

Shoudl the need to be added as well?

coryan replied:
No, those are for backwards compatiblity.

jsrinnn replied:
Done.

jsrinnn commented:
Pushed latest review feedback changes. PTAL.

Thanks!

coryan reviewed (commented):
This is getting pretty close.

The robots are unhappy about the CMake file formatting.  Running `checkers` should fix that.

Annoyingly, one of our builds detects that only expected directories are created during the install and needs an update.  Add the new directories here:

https://github.com/googleapis/google-cloud-cpp/blob/3757a3647ade6a491803c493fe6484285de73526/ci/cloudbuild/builds/cmake-install.sh#L62-L64

jsrinnn commented:
Thanks Carlos. Will do. I am almost done verifying the build changes. Took a while becaus I was running both cmake build and bazel build and integration tests for verification. I am almost done with that.

I will add the new directory to the install file you suggest. 

A followup question for cmake formatting. Is there a comamnd I can run for that to fix any format errors i.e something equivalent to "g4 fix" in google-3?

coryan commented:
> A followup question for cmake formatting. Is there a comamnd I can run for that to fix any format errors i.e something equivalent to "g4 fix" in google-3?

I use:

```
ci/cloudbuild/build.sh -t checkers-pr
```

An important difference with `g4 fix` is that this will change the files *locally*.  You still need to commit the changes it makes and push them to the repository.

jsrinnn commented:
Thank you. Followup question regarding the command below

ci/cloudbuild/build.sh -t checkers-pr

Does this only look at commited/pushed files or can it be used for untracked files as well?

coryan commented:
> Does this only look at commited/pushed files or can it be used for untracked files as well?

Somewhere in between those two.  It only looks at files returned by `git ls-files`.  So it would ignore untracked files, but it will include staged files (e.g. files that are `git add`-ed, but not committed yet), as well as committed files.

jsrinnn commented:
I have pushed the following changes
1) Added GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN and GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END macros to all the files
2) Modified Cmake and Bazel build files to include the files in v2/minimal/internal
3) Fixed formatting errors and added the v2/minimal/internal to cmake-install
4) Merged latest sync from upsteam main

I ran all both Cmake and Bazel builds and also. ran the integration tests.

PTAL. Thank you!

jsrinnn commented:
Carlos,

Looks like cmake-install-pr is failing. Is there anywhere else I need to add the new directory? Thanks

coryan commented on ci/cloudbuild/builds/cmake-install.sh:
```diff
@@ -61,6 +61,7 @@ expected_dirs+=(
   ./include/google/appengine/logging/v1
   ./include/google/cloud/bigquery/logging
   ./include/google/cloud/bigquery/logging/v1
+  ./include/google/cloud/bigquery/v2/minimal/internal
```
This is missing `./include/google/cloud/bigquery/v2/minimal`

jsrinnn replied:
Do I need to add the parent directory ./include/google/cloud/bigquery/v2 as well?

coryan replied:
Yes, the script is too dumb to figure it out.

coryan replied:
This build is getting on my nerves...  It does not handle duplicates and `.../v2` was already listed here:

https://github.com/googleapis/google-cloud-cpp/blob/3757a3647ade6a491803c493fe6484285de73526/ci/etc/expected_install_directories#L18

jsrinnn replied:
Ah...alright. Once more unto... :-)

coryan replied:
Quoting Henry V just made my day

coryan reviewed (approved):
LGTM, please fix the build problem and then you should be able to merge.

jsrinnn commented:
Pushed the cmake-install changes to add the parent directories

jsrinnn commented:
cmake-install is still failing :(. Anything obvious that I am missing?

jsrinnn commented:
Ah...looks like I may need to add the files here as well:
https://github.com/googleapis/google-cloud-cpp/blob/ff7575357ce91526a1bf4d271bc6a0ef9bf949cd/ci/etc/expected_install_directories

jsrinnn commented:
Latest push removes the duplicate bigquery/v2. Hopefully this fixes it.

jsrinnn commented:
Merging the initial files for V2 Job resource along with Cmake and Bazel build changes to include the new files.

For detailed design please see here: http://goto.google.com/bq-odbc-client-design
